ffmpeg_new.c

```diff
@@ -0,0 +1,555 @@
+	avformat_close_input(&priv->input_context);
+	swr_free(&priv->swr);
+	ffmpeg_input_free(priv->input);
+	ffmpeg_output_free(priv->output);
+	free(priv);
+	ip_data->private = NULL;
+	return 0;
+}
+
+static int ffmpeg_fill_buffer(AVFormatContext *ic, AVCodecContext *cc,
+		struct ffmpeg_input *input, struct ffmpeg_output *output, SwrContext *swr)
+{
+	AVFrame *frame = ffmpeg_frame_alloc();
+	int res;
+
+again:
+	if (!ffmpeg_receive_frame(cc, frame)) {
+		res = swr_convert(swr,
+				  &output->buffer,
+				  frame->nb_samples,
+				  (const uint8_t **)frame->extended_data,
+				  frame->nb_samples);
+		if (res < 0) {
+			res = -IP_ERROR_INTERNAL;
+			goto out;
+		}
+		output->buffer_pos = output->buffer;
+		output->buffer_used_len = res * cc->channels * sizeof(int16_t);
+		res = output->buffer_used_len;
+		goto out;
+	}
+
+	if (input->eof) {
+		res = 0;
+		goto out;
+	}
+
+	ffmpeg_packet_unref(&input->pkt);
+	res = av_read_frame(ic, &input->pkt);
+
+	if (res < 0) {
+		if (res == AVERROR_EOF) {
+			ffmpeg_send_packet(cc, NULL);
+			input->eof = true;
+			goto again;
+		} else {
+			res = 0;
+			goto out;
+		}
+	}
+
+	if (input->pkt.stream_index == input->stream_index) {
+		input->curr_size += input->pkt.size;
+		input->curr_duration += input->pkt.duration;
+		if (ffmpeg_send_packet(cc, &input->pkt)) {
+			res = -IP_ERROR_INTERNAL;
+			goto out;
+		}
+	}
+
+	goto again;
+
+out:
+	ffmpeg_frame_free(&frame);
+	return res;
+}
+
+static int ffmpeg_read(struct input_plugin_data *ip_data, char *buffer, int count)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	struct ffmpeg_output *output = priv->output;
+	int rc;
+	int out_size;
+
+	if (output->buffer_used_len == 0) {
+		rc = ffmpeg_fill_buffer(priv->input_context, priv->codec_context,
+				priv->input, priv->output, priv->swr);
+		if (rc <= 0) {
+			return rc;
+		}
+	}
+	out_size = min(output->buffer_used_len, count);
+	memcpy(buffer, output->buffer_pos, out_size);
+	output->buffer_used_len -= out_size;
+	output->buffer_pos += out_size;
+	return out_size;
+}
+
+static int ffmpeg_seek(struct input_plugin_data *ip_data, double offset)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	AVStream *st = priv->input_context->streams[priv->input->stream_index];
+	int ret;
+
+	int64_t pts = av_rescale_q(offset * AV_TIME_BASE, AV_TIME_BASE_Q, st->time_base);
+
+	avcodec_flush_buffers(priv->codec_context);
+	/* Force reading a new packet in next ffmpeg_fill_buffer(). */
+	priv->input->curr_pkt_size = 0;
+
+	ret = av_seek_frame(priv->input_context, priv->input->stream_index, pts, 0);
+
+	if (ret < 0) {
+		return -IP_ERROR_FUNCTION_NOT_SUPPORTED;
+	} else {
+		ffmpeg_buffer_flush(priv->output);
+		return 0;
+	}
+}
+
+static void ffmpeg_read_metadata(struct growing_keyvals *c, AVDictionary *metadata)
+{
+	AVDictionaryEntry *tag = NULL;
+
+	while ((tag = av_dict_get(metadata, "", tag, AV_DICT_IGNORE_SUFFIX))) {
+		if (tag->value[0])
+			comments_add_const(c, tag->key, tag->value);
```

Comment:
Yeah — I was just reading the code more from the «one plugin to rule them all» than «updated ffmpeg plugin» point of view.